New report API

pjanik opened the pull request:
`GET api/v1/reports/offering?id=:id`
Example response: https://gist.github.com/pjanik/b831b67f659e709931bf

It provides some basic information, students list and finally the report structure. It's a tree: (investigation) -> activities -> sections -> pages -> questions -> answers. If offering is an activity, the root will be different (activity). It includes filters settings too (`visible_in_report`).

We don't necessarily need to merge it very soon, things might change a few times once I start working on the client side. However some initial feedback, especially regarding structure and general approach, would be useful. 

@scytacki, @knowuh, a few general notes following our discussion on Monday:
I decided not to use Doug's `api/v1/answer_controller`. It provides students answer, but the format wouldn't be really useful for me. Also, it seems to have lots of raw SQL, maybe it makes sense due to performance reasons in this case (as it provides way more data than my single report), but it wouldn't be too easy to customize it. Also, obtaining students answer was actually the easiest part of this work (just read `Report::Learners`). 

I was trying to follow what's going on when we render the current report and base implementation on it, so hopefully things still work in the same way. The old controller itself is actually very simple, but it passes a bunch of objects to view and then the craziness begins. Views call helpers, partials, then some models, then you go back to view, plus there are some weird intermediate objects (Saveables, Standins) and so on. I tried to simplify it as much as I could, provide minimum information which is necessary to generate current report and move rest of the logic to React components. Yes, I might have missed something important, but I feel it's better to test it in practice rather than try to untangle the old code. Once we're sure that it's working, I hope we can get rid of many files, e.g. [report/util.js](https://github.com/concord-consortium/rigse/blob/master/app/models/report/util.rb). I feel the old code is overly complicated as it randomly tries to cache and pre-calculate various things, so views don't hit DB too many times. JSON structure should solve that issue. I tried not to use old helpers at all, even if some methods were useful, and to keep all the logic in this one controller. If it works and we could limit report generation to < 200 lines, it would be great.

Performance should also be way better, probably we can even implement live reload of the report (easy with React).

knowuh commented on app/controllers/api/v1/reports_controller.rb:
```diff
@@ -0,0 +1,190 @@
+    if ans[:type] === 'Dataservice::Blob'
+      ans[:image_url] = dataservice_blob_raw_url(id: ans[:id], token: ans[:token])
+      # Skip unnecessary attributes, leave only image url and note.
+      ans.slice!(:image_url, :note)
```
Wow. I didn't know about Hash#slice.  ❤

knowuh commented:
This looks good.  Very clean and simple.

Putting all these private helper methods in the controller makes it easy to review because its all in one file. But It does make it hard to test the serialized form of the json of component parts.  Perhaps we are not interested in testing those bits?

pjanik commented:
> Putting all these private helper methods in the controller makes it easy to review because its all in one file. But It does make it hard to test the serialized form of the json of component parts. Perhaps we are not interested in testing those bits?

Yes, good point, I was initially planning to put all those methods in a separate module, just as a set of static functions, so it's easy to test. But then I had to use `_url` func, so I would need to pass host name and port all the way down in the call chain, so it was becoming strange. I'll think about that for sure, especially that the controller will grow soon (filters, feedback). I didn't focus on testing at the moment, as I think this format may change 10 times once we have client-side code.

On the other hand, we might just prepare some fake data and simply check if API returns expected JSON. Simple testing against regression and unwanted API changes. In this case, I guess probably we want to check the whole answer anyway.

scytacki commented:
Going off of what @knowuh said, I think it would be good to pull the private methods into an independent class. The controller can instantiate it with the offering and the host and then you wouldn't need to pass the offering and host around to each of the methods.  Then you can test that object without having to worry about the current_user.

pjanik commented:
I wanted to avoid creating classes, state, etc., and just keep these methods pure, `(input_model) -> output.json`, so it's easy to test them (maybe too much reading about react, redux, etc.?). But in this case it's either annoying (number of arguments) or requires (too) fancy ruby techniques. I'll change that to simple helper class.

pjanik commented:
I've moved helpers to a separate class and added new route, so filter can be updated.

knowuh commented on app/controllers/api/v1/reports_controller.rb:
```diff
@@ -0,0 +1,38 @@
+  def update
+    offering = Portal::Offering.find(params[:id])
+    # authorize offering, :api_report?
+    puts params
```
Forgotten debug statement

knowuh commented on app/controllers/api/v1/reports_controller.rb:
```diff
@@ -0,0 +1,38 @@
+  private
+
+  def update_visibility_filter(filter, filter_params)
+    filter.embeddable_keys = filter_params[:questions] unless filter_params[:questions].nil?
```
To unset the filter, I guess we need to pass in empty array.

pjanik replied:
Do mean the explicit `.nil?` check might be confusing (as `present?` / `blank?` would fail here for empty array edge-case)?

Otherwise I'm not sure if I understand. Basically the filter consists of list of questions + enabled / disabled flag (named `ignore`). They can be set independent, so you can keep questions selected, but still deactivate the filter. And this API seems to reflect that, just sets two properties + makes sure we won't unset the second property if only one is provided.

knowuh commented on app/models/api/v1/report.rb:
```diff
@@ -0,0 +1,226 @@
+  end
+
+  def process_image_question_answer(hash)
+    # Filter out useless Blob description, provide image URL instead (and note / annotation).
```
I feel like we should eventually push blob reformatting deeper into the reporting stack.  I was doing something similar with the feedback work.  I might come back to look at this.

knowuh commented:
@pjanik  This looks great.  :+1:   (there's a puts statement which probably isn't needed any more)

scytacki commented:
We need to enable the authorize statements before merging.

pjanik commented:
Thanks @knowuh, I've removed `put` and added some comments. 
@scytacki, I've enabled authorization and added auth support in the client side app (+ better error handling). It seems to be working fine on both sides (I've extracted token from my DB and pasted it to the URL manually). So, the only missing thing is launch URL generation, but I guess it'll be pretty much the same like for the HAS Dashboard.

* Wrong token: ![screen shot 2016-03-10 at 14 03 48](https://cloud.githubusercontent.com/assets/767857/13670251/13bc83a6-e6c9-11e5-83bd-a58dd85d4804.png)

* Working token: ![screen shot 2016-03-10 at 14 03 24](https://cloud.githubusercontent.com/assets/767857/13670230/f93d89c6-e6c8-11e5-81f3-c00cdc225ac7.png)

scytacki commented:
You can also test it using the External Report system in the portal.  I'll try that out in shortly :)

:+1:

scytacki commented:
Oh I guess the external report system can't be used because the reportUrl argument needs to be dynamically set, and the external report system uses a different reportUrl. It would be good to unify that if possible.
